controller/src/beerocks/master/wfa_ca.cpp

```diff
@@ -844,6 +889,92 @@ bool wfa_ca::get_send_1905_1_tlv_hex_list(std::list<tlv_hex_t> &tlv_hex_list,
     return false;
 }
 
+/**
+ * @brief Writes hex string into the class buffer.
+ * 
+ * @param[in] value string in hex notation.
+ * @param[out] length Length of current tlv.
+ * @return false if length smaller than data, else true.
+ */
+bool tlvPrefilledData::add_tlv_value_hex_string(const std::string &value, uint16_t &length)
+{
+    // iterate every two chars (1 octet) and write it to buffer
+    // start with char_idx = 2 to skip "0x"
+    for (size_t char_idx = 2; char_idx < value.size(); char_idx += 2) {
+        if (length == 0) {
+            return false;
+        }
+        *m_buff_ptr__ = std::stoi(value.substr(char_idx, 2), nullptr, 16);
+        m_buff_ptr__++;
+        length--;
+    }
+    return true;
+}
+
+/**
+ * @brief Writes a decimal string into the class buffer.
+ * 
+ * @param[in] value Decimal string.
+ * @param[out] length Length of current tlv.
+ * @return false if length smaller than data, else true.
+ */
+bool tlvPrefilledData::add_tlv_value_decimal_string(const std::string &value, uint16_t &length)
+{
+    int i_dec        = std::stoi(value.substr(2, value.length()));
+    int num_of_bytes = i_dec / UINT8_MAX + (i_dec % UINT8_MAX != 0 ? 1 : 0);
```

Comment:
Sorry, @morantr, but this is incorrect... First of all, UINT8_MAX is 255, not 256. More importantly, though, we need log<sub>256</sub>(i_dec) but you're calculating i_dec/256 rounded up...

That said, as I mentioned in previous reviews, this `0d` stuff is utterly broken anyway. So I think we could make the assumption that it's either 1 or 2 bytes. Also, I think we should do it based on the number of digits. So:
```suggestion
    // Assume 2 bytes if it has more than 3 digits
    if (value.size() > 5) {
        *m_buff_ptr__ = i_dec / 256;
        m_buff_ptr__++;
        length--;
    }
    *m_buff_ptr__ = i_dec % 256;
    m_buff_ptr__++;
    length--;
```

This is based on the example that @CoralMalachi found of a `0d0600` value in one of the test cases. Note that I don't really like relying on the i_dec value anyway, because it is possible that they use something like `0d0005` string somewhere for a two-byte value...

But that means, of course, that if we ever find any sample where `0d` is used for 3- or 4-byte values, we have to change this again... So to be a bit more prepared for that, we could still set `num_of_bytes` (to one or two, depending on `value.size()`) and keep the loop.

Reply:
@arnout, yes I can see now that what I suggested will work only on decimals between 0 - 509
:(